### ISSUE TITLE
Font cache databe new features request

### ISSUE THREAD

**Issue by linasstonys** [latex3/luaotfload#151]
I want ask for two new features in font cache system (`luaotfload-database.lua`): 

1. Possibility to forbid system fonts ( bool variable `config.luaotfload.db.scan_system`) - the problem is if several people work with the same tex file from different OS. And if system font will be used this tex file can't be compiled in other machine (if there is no such font). If you could add a swich like `config.luaotfload.db.scan_system` like you did with `config.luaotfload.db.scan_local` it would fix this problem.

```
local collect_font_filenames = function ()
    ...
    if config.luaotfload.db.scan_system == true then               -- HERE
        tableappend (filenames, collect_font_filenames_system ())
    end                                                            -- HERE
    if config.luaotfload.db.scan_local == true then
        tableappend (filenames, collect_font_filenames_local  ())
    end
    ...
end
```

2) There is a problem with absolute file names from the current working directory (font is in some local directory) if you copy that directory to other place. For example you have a dockerized tex compilation system which copies files in randomly created temp directory, recompiles files and returns all files set. First time compiling it will create font cache locally (TEXMFCACHE=.) and on an other compilation there will be an error because `luaotfload-names.lua.gz` has stored font from temporary directory with an absolute path.
The fix could be to add a new bool switch `config.luaotfload.db.use_relpath` where PWD would be replaced by '.'. Something similar like:

```
local collect_font_filenames_dir = function (dirname, location)
    ...
    local pwd = false
    if config.luaotfload.db.use_relpath then
        pwd = path_normalize (lpegmatch (stripslashes, lfscurrentdir ()))
    end
    
    for j = 1, nfound do
        local fullname = found[j]
        local normalized_path = path_normalize (fullname)
        if config.luaotfload.db.use_relpath then        
            if string.find(normalized_path, '^'..pwd) ~= nil then
                normalized_path = "." .. string.sub(normalized_path, #pwd + 1)
            end
        end
        files[#files + 1] = { normalized_path, location }
    end
    ...
end
```

It would be very nice to have these features in the future. Now I just modify `luaotfload-database.lua` and put it to every texlive locally.

**Comment by zauguin** [comment]
> I want ask for two new features in font cache system (`luaotfload-database.lua`):
> 
>     1. Possibility to forbid system fonts ( bool variable `config.luaotfload.db.scan_system`) - the problem is if several people work with the same tex file from different OS. And if system font will be used this tex file can't be compiled in other machine (if there is no such font). If you could add a swich like `config.luaotfload.db.scan_system` like you did with `config.luaotfload.db.scan_local` it would fix this problem.

Of course, this would require everyone involved to actually set the switch then and it would affect all documents. Also everyone would have to rebuild their database to change this setting.

On the other hand, there already is a nice feature for these cases which doesn't have these problems: Use `kpse:` lookups which never look into the system directories in the first place and it can be used individually for every font without having to modify system settings.

Anyway I am currently looking into adding a configuration file interface for `fonts.names.set_location_precedence`. That would also give a solution for this because the precedence order can be changed to not include system fonts. (The primary issue with all these is Windows: There the distinction between system and texmf fonts is a bit fuzzy.)

> 
> 
> ```
> local collect_font_filenames = function ()
>     ...
>     if config.luaotfload.db.scan_system == true then               -- HERE
>         tableappend (filenames, collect_font_filenames_system ())
>     end                                                            -- HERE
>     if config.luaotfload.db.scan_local == true then
>         tableappend (filenames, collect_font_filenames_local  ())
>     end
>     ...
> end
> ```
> 
>     1. There is a problem with absolute file names from the current working directory (font is in some local directory) if you copy that directory to other place. For example you have a dockerized tex compilation system which copies files in randomly created temp directory, recompiles files and returns all files set. First time compiling it will create font cache locally (TEXMFCACHE=.) and on an other compilation there will be an error because `luaotfload-names.lua.gz` has stored font from temporary directory with an absolute path.
>        The fix could be to add a new bool switch `config.luaotfload.db.use_relpath` where PWD would be replaced by '.'. Something similar like:
> 

Actually there are two parts of this: First there is a bug in the filename lookup logic. It can not deal with the event that font which is loaded using it's filename does not actually exists. This leads to problems whenever people try to move or delete fonts. That bug should be fixed in the `dev` branch with bd1358ef.

After this fix, your setup should work. The remaining issue is that it keeps updating the cache and therefore wasting some time everytime you move around your fonts. I'm not yet convinced that this is worth "fixing" though because it is pretty specific to this situation and we really want to avoid anything involving relative paths in the regular situation where TEXMFCACHE is in a fixed location.

Also on an off-topic note, what is the motivation for using randomized locations anyway? I haven't used docker in a while, but doesn't it basically provide an isolated, virtualized filesystem anyway?

**Comment by linasstonys** [comment]
"Also on an off-topic note, what is the motivation for using randomized locations anyway? I haven't used docker in a while, but doesn't it basically provide an isolated, virtualized filesystem anyway?"
